File: apps/mobile/lib/titan/core/surface_contract.dart
/// Flutter mirror of packages/titan-platform/src/surface/index.ts.
/// Keep this DTO contract schema-compatible with Titan Core; it grants no authority.
enum TitanSurface { zero, go, hub }

TitanSurface titanSurfaceFromWire(String value) {
  switch (value) {
    case 'zero':
    case 'command':
      return TitanSurface.zero;
    case 'go':
      return TitanSurface.go;
    case 'hub':
      return TitanSurface.hub;
    default:
      throw FormatException('canonical-surface-required');
  }
}

String titanSurfaceToWire(TitanSurface surface) => switch (surface) {
  TitanSurface.zero => 'zero',
  TitanSurface.go => 'go',
  TitanSurface.hub => 'hub',
};

class TitanSurfaceCapability {
  final String capabilityId;
  final List<String> operations;
  final bool mutation;
  final String offline;
  final bool requiresReceipt;
  final String? authorityCeiling;
  const TitanSurfaceCapability({required this.capabilityId, required this.operations, required this.mutation, required this.offline, required this.requiresReceipt, this.authorityCeiling});
  factory TitanSurfaceCapability.fromJson(Map<String,dynamic> json) => TitanSurfaceCapability(
    capabilityId: json['capability_id'] as String,
    operations: List<String>.from(json['operations'] as List),
    mutation: json['mutation'] == true,
    offline: (json['offline'] as String?) ?? 'read',
    requiresReceipt: json['requires_receipt'] == true,
    authorityCeiling: json['authority_ceiling'] as String?,
  );
}

class TitanSurfaceProjection {
  final String companyId;
  final TitanSurface surface;
  final String actorId;
  final String revision;
  final DateTime issuedAt;
  final DateTime expiresAt;
  final List<TitanSurfaceCapability> capabilities;
  final Map<String,dynamic> data;
  const TitanSurfaceProjection({required this.companyId, required this.surface, required this.actorId, required this.revision, required this.issuedAt, required this.expiresAt, required this.capabilities, required this.data});
  factory TitanSurfaceProjection.fromJson(Map<String,dynamic> json) {
    if (json.containsKey('tenant_company_id') || json.containsKey('tenant_id')) throw const FormatException('tenant_company_id-not-authoritative');
    final companyId=(json['company_id'] as String?)?.trim() ?? '';
    if (companyId.isEmpty) throw const FormatException('company_id-required');
    if (json['authority_neutral'] != true || json['identity_grants_authority'] != false || json['cached_state_grants_authority'] != false) throw const FormatException('surface-authority-contract-invalid');
    return TitanSurfaceProjection(
      companyId:companyId,
      surface:titanSurfaceFromWire(json['surface'] as String),
      actorId:json['actor_id'] as String,
      revision:json['revision'] as String,
      issuedAt:DateTime.parse(json['issued_at'] as String),
      expiresAt:DateTime.parse(json['expires_at'] as String),
      capabilities:(json['capabilities'] as List).map((e)=>TitanSurfaceCapability.fromJson(Map<String,dynamic>.from(e as Map))).toList(growable:false),
      data:Map<String,dynamic>.from((json['data'] as Map?) ?? const {}),
    );
  }
  TitanSurfaceCapability requireCapability(String id,String operation,{DateTime? now}) {
    if (!expiresAt.isAfter(now ?? DateTime.now().toUtc())) throw StateError('surface-projection-expired');
    return capabilities.firstWhere((c)=>c.capabilityId==id && c.operations.contains(operation),orElse:()=>throw StateError('surface-capability-not-authorised'));
  }
}

File: apps/web/app/titan/runtime/surface-contract.mjs
export const SURFACE_CONTRACT_VERSION = "1.0";

const presentation = Object.freeze({
  go: Object.freeze({
    name: "Titan Go",
    audience: "Field staff",
    accent: "#2563eb",
    heading: "What do you need next?",
    greeting: "Morning, Alex. Your first job is ready. The Access Failure Prevention Specialist has confirmed entry and the route is clear.",
    prompt: "Ask about today’s work, a customer, or what to do next…",
    suggestions: Object.freeze(["What’s my next job?", "Start Hartwell Dental", "Report issue"]),
  }),
  hub: Object.freeze({
    name: "Titan Hub",
    audience: "Customers",
    accent: "#16a34a",
    heading: "How can I help with your service?",
    greeting: "Hi Sam. Your service is confirmed for tomorrow at 9:00 AM and arrival updates are on.",
    prompt: "Ask about a booking, quote, invoice, or your service…",
    suggestions: Object.freeze(["Track my service team", "Reschedule tomorrow", "Show service add-ons"]),
  }),
  zero: Object.freeze({
    name: "Titan Command",
    audience: "Owners & managers",
    accent: "#d97706",
    heading: "What needs your decision?",
    greeting: "Good evening, Jason. Your business is healthy. Two decisions need you.",
    prompt: "Ask anything about your business…",
    suggestions: Object.freeze(["What needs me now?", "Show today’s operations", "What can the AI workforce handle?"]),
  }),
});

const capabilities = Object.freeze({
  go: Object.freeze([
    { capability_id: "jobs.read", operations: ["read"], mutation: false, offline: "read" },
    { capability_id: "jobs.progress", operations: ["travel", "arrive", "start", "complete"], mutation: true, offline: "queue", requires_receipt: true },
    { capability_id: "messages.send", operations: ["send"], mutation: true, offline: "queue", requires_receipt: true },
    { capability_id: "evidence.capture", operations: ["prepare", "upload"], mutation: true, offline: "queue", requires_receipt: true },
    { capability_id: "maps.navigate", operations: ["read", "navigate"], mutation: false, offline: "read" },
  ]),
  hub: Object.freeze([
    { capability_id: "services.read", operations: ["read"], mutation: false, offline: "read" },
    { capability_id: "services.reschedule", operations: ["prepare", "request"], mutation: true, offline: "forbidden", requires_receipt: true },
    { capability_id: "messages.send", operations: ["send"], mutation: true, offline: "queue", requires_receipt: true },
    { capability_id: "payments.intent", operations: ["prepare", "submit"], mutation: true, offline: "forbidden", requires_receipt: true },
    { capability_id: "addons.request", operations: ["prepare", "submit"], mutation: true, offline: "forbidden", requires_receipt: true },
  ]),
  zero: Object.freeze([
    { capability_id: "decisions.read", operations: ["read"], mutation: false, offline: "read" },
    { capability_id: "decision.resolve", operations: ["approve", "deny"], mutation: true, offline: "forbidden", requires_receipt: true },
    { capability_id: "operations.read", operations: ["read"], mutation: false, offline: "read" },
    { capability_id: "workforce.read", operations: ["read"], mutation: false, offline: "read" },
    { capability_id: "workforce.control", operations: ["pause", "resume", "adjust"], mutation: true, offline: "forbidden", requires_receipt: true },
    { capability_id: "system.read", operations: ["read"], mutation: false, offline: "read" },
    { capability_id: "rewind.preview", operations: ["read", "preview"], mutation: false, offline: "read" },
  ]),
});

function required(value, label) {
  if (typeof value !== "string" || value.trim() === "") throw new TypeError(`${label}-required`);
  return value.trim();
}

function rejectLegacy(input) {
  if (input && typeof input === "object" && ("tenant_company_id" in input || "tenant_id" in input)) {
    throw new TypeError("tenant_company_id-not-authoritative");
  }
}

export function createSurfaceProjection(input) {
  rejectLegacy(input);
  const surface = required(input?.surface, "canonical-surface");
  if (!(surface in presentation)) throw new TypeError("canonical-surface-required");
  const company_id = required(input?.company_id, "company-id");
  const actor_id = required(input?.actor_id, "actor-id");
  const revision = required(input?.revision, "projection-revision");
  const issued_at = required(input?.issued_at, "projection-issued-at");
  const expires_at = required(input?.expires_at, "projection-expires-at");
  if (!Number.isFinite(Date.parse(issued_at))) throw new TypeError("projection-issued-at-invalid");
  if (!Number.isFinite(Date.parse(expires_at))) throw new TypeError("projection-expires-at-invalid");
  return Object.freeze({
    schema_version: SURFACE_CONTRACT_VERSION,
    company_id,
    surface,
    actor_id,
    revision,
    issued_at,
    expires_at,
    capabilities: capabilities[surface].map((capability) => Object.freeze({
      ...capability,
      operations: Object.freeze([...capability.operations]),
      requires_receipt: capability.requires_receipt ?? capability.mutation,
    })),
    data: Object.freeze({ presentation: presentation[surface] }),
    authority_neutral: true,
    identity_grants_authority: false,
    cached_state_grants_authority: false,
  });
}

export function getDemoSurfaceProjection(surface) {
  if (!(surface in presentation)) throw new TypeError("canonical-surface-required");
  return createSurfaceProjection({
    company_id: "demo_001",
    surface,
    actor_id: `${surface}-demo-actor`,
    revision: "merge77-demo-rev-1",
    issued_at: "2026-09-16T04:00:00.000Z",
    expires_at: "2099-09-16T04:00:00.000Z",
  });
}

export function createSurfaceCommandIntent(input) {
  rejectLegacy(input);
  const capability = input.projection.capabilities.find((entry) =>
    entry.capability_id === input.capability_id && entry.operations.includes(input.operation),
  );
  if (!capability) throw new TypeError("surface-capability-not-authorised");
  if (Date.parse(input.projection.expires_at) <= Date.now()) throw new TypeError("surface-projection-expired");
  return Object.freeze({
    schema_version: SURFACE_CONTRACT_VERSION,
    command_id: input.command_id ?? `demo-command-${Date.now()}`,
    company_id: input.projection.company_id,
    surface: input.projection.surface,
    actor_id: input.projection.actor_id,
    projection_revision: input.projection.revision,
    capability_id: required(input.capability_id, "surface-capability-id"),
    operation: required(input.operation, "surface-operation"),
    idempotency_key: required(input.idempotency_key, "idempotency-key"),
    correlation_id: required(input.correlation_id, "correlation-id"),
    payload: Object.freeze({ ...(input.payload ?? {}) }),
    transport: "titan-command-bus",
    execution_authorised: false,
    requires_server_acceptance: true,
    requires_receipt: capability.requires_receipt,
  });
}
